File: ja/codes/python/chapter_tree/array_binary_tree.py
"""
File: array_binary_tree.py
Created Time: 2023-07-19
Author: krahets (krahets@163.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import TreeNode, list_to_tree, print_tree


class ArrayBinaryTree:
    """配列表現による二分木クラス"""

    def __init__(self, arr: list[int | None]):
        """コンストラクタ"""
        self._tree = list(arr)

    def size(self):
        """リスト容量"""
        return len(self._tree)

    def val(self, i: int) -> int | None:
        """インデックス i のノードの値を取得"""
        # インデックスが範囲外なら、空きを表す None を返す
        if i < 0 or i >= self.size():
            return None
        return self._tree[i]

    def left(self, i: int) -> int | None:
        """インデックス i のノードの左子ノードのインデックスを取得"""
        return 2 * i + 1

    def right(self, i: int) -> int | None:
        """インデックス i のノードの右子ノードのインデックスを取得"""
        return 2 * i + 2

    def parent(self, i: int) -> int | None:
        """インデックス i のノードの親ノードのインデックスを取得"""
        return (i - 1) // 2

    def level_order(self) -> list[int]:
        """レベル順走査"""
        self.res = []
        # 配列を直接走査する
        for i in range(self.size()):
            if self.val(i) is not None:
                self.res.append(self.val(i))
        return self.res

    def dfs(self, i: int, order: str):
        """深さ優先探索"""
        if self.val(i) is None:
            return
        # 先行順走査
        if order == "pre":
            self.res.append(self.val(i))
        self.dfs(self.left(i), order)
        # 中順走査
        if order == "in":
            self.res.append(self.val(i))
        self.dfs(self.right(i), order)
        # 後順走査
        if order == "post":
            self.res.append(self.val(i))

    def pre_order(self) -> list[int]:
        """先行順走査"""
        self.res = []
        self.dfs(0, order="pre")
        return self.res

    def in_order(self) -> list[int]:
        """中順走査"""
        self.res = []
        self.dfs(0, order="in")
        return self.res

    def post_order(self) -> list[int]:
        """後順走査"""
        self.res = []
        self.dfs(0, order="post")
        return self.res


"""Driver Code"""
if __name__ == "__main__":
    # 二分木を初期化
    # ここでは、配列から直接二分木を生成する関数を利用する
    arr = [1, 2, 3, 4, None, 6, 7, 8, 9, None, None, 12, None, None, 15]
    root = list_to_tree(arr)
    print("\n二分木を初期化\n")
    print("二分木の配列表現：")
    print(arr)
    print("二分木の連結リスト表現：")
    print_tree(root)

    # 配列表現による二分木クラス
    abt = ArrayBinaryTree(arr)

    # ノードにアクセス
    i = 1
    l, r, p = abt.left(i), abt.right(i), abt.parent(i)
    print(f"\n現在のノードのインデックスは {i}、値は {abt.val(i)}")
    print(f"その左子ノードのインデックスは {l}、値は {abt.val(l)}")
    print(f"その右子ノードのインデックスは {r}、値は {abt.val(r)}")
    print(f"その親ノードのインデックスは {p}、値は {abt.val(p)}")

    # 木を走査
    res = abt.level_order()
    print("\nレベル順走査：", res)
    res = abt.pre_order()
    print("先行順走査：", res)
    res = abt.in_order()
    print("中間順走査：", res)
    res = abt.post_order()
    print("後行順走査：", res)

File: ja/codes/javascript/chapter_tree/array_binary_tree.js
/**
 * File: array_binary_tree.js
 * Created Time: 2023-08-06
 * Author: yuan0221 (yl1452491917@gmail.com)
 */

const { arrToTree } = require('../modules/TreeNode');
const { printTree } = require('../modules/PrintUtil');

/* 配列表現による二分木クラス */
class ArrayBinaryTree {
    #tree;

    /* コンストラクタ */
    constructor(arr) {
        this.#tree = arr;
    }

    /* リスト容量 */
    size() {
        return this.#tree.length;
    }

    /* インデックス i のノードの値を取得 */
    val(i) {
        // インデックスが範囲外なら、空きを表す null を返す
        if (i < 0 || i >= this.size()) return null;
        return this.#tree[i];
    }

    /* インデックス i のノードの左子ノードのインデックスを取得 */
    left(i) {
        return 2 * i + 1;
    }

    /* インデックス i のノードの右子ノードのインデックスを取得 */
    right(i) {
        return 2 * i + 2;
    }

    /* インデックス i のノードの親ノードのインデックスを取得 */
    parent(i) {
        return Math.floor((i - 1) / 2); // 切り捨て除算
    }

    /* レベル順走査 */
    levelOrder() {
        let res = [];
        // 配列を直接走査する
        for (let i = 0; i < this.size(); i++) {
            if (this.val(i) !== null) res.push(this.val(i));
        }
        return res;
    }

    /* 深さ優先探索 */
    #dfs(i, order, res) {
        // 空きスロットなら返す
        if (this.val(i) === null) return;
        // 先行順走査
        if (order === 'pre') res.push(this.val(i));
        this.#dfs(this.left(i), order, res);
        // 中順走査
        if (order === 'in') res.push(this.val(i));
        this.#dfs(this.right(i), order, res);
        // 後順走査
        if (order === 'post') res.push(this.val(i));
    }

    /* 先行順走査 */
    preOrder() {
        const res = [];
        this.#dfs(0, 'pre', res);
        return res;
    }

    /* 中順走査 */
    inOrder() {
        const res = [];
        this.#dfs(0, 'in', res);
        return res;
    }

    /* 後順走査 */
    postOrder() {
        const res = [];
        this.#dfs(0, 'post', res);
        return res;
    }
}

/* Driver Code */
// 二分木を初期化
// ここでは、配列から直接二分木を生成する関数を利用する
const arr = Array.of(
    1,
    2,
    3,
    4,
    null,
    6,
    7,
    8,
    9,
    null,
    null,
    12,
    null,
    null,
    15
);

const root = arrToTree(arr);
console.log('\n二分木を初期化\n');
console.log('二分木の配列表現：');
console.log(arr);
console.log('二分木の連結リスト表現：');
printTree(root);

// 配列表現による二分木クラス
const abt = new ArrayBinaryTree(arr);

// ノードにアクセス
const i = 1;
const l = abt.left(i);
const r = abt.right(i);
const p = abt.parent(i);
console.log('\n現在のノードのインデックスは ' + i + ' ，値は ' + abt.val(i));
console.log(
    'その左子ノードのインデックスは ' + l + ' ，値は ' + (l === null ? 'null' : abt.val(l))
);
console.log(
    'その右子ノードのインデックスは ' + r + ' ，値は ' + (r === null ? 'null' : abt.val(r))
);
console.log(
    'その親ノードのインデックスは ' + p + ' ，値は ' + (p === null ? 'null' : abt.val(p))
);

// 木を走査
let res = abt.levelOrder();
console.log('\nレベル順走査：' + res);
res = abt.preOrder();
console.log('先行順走査：' + res);
res = abt.inOrder();
console.log('中間順走査：' + res);
res = abt.postOrder();
console.log('後行順走査：' + res);
